### ISSUE TITLE
`begin ... end` expression

### ISSUE THREAD

**Issue by rhysd** [rhysd/Dachs#46]
Sometimes I use lambda expression which is invocated instantly because I want to consume some statements only for initialization of variable and make the variable immutable.

```
p := 42
q := 12

x := -> do
    a := some_process(p, q)
    if a > 0
        ret p * a
    else
        ret some_process(p, a)
    end
end()
```

This is better than below

```
p := 42
q := 12

var x : int
do
    a := some_process(p, q)
    if a > 0
        x = p * a
    else
        x = some_process(p, a)
    end
end
```

because (1) `x` must be default constructible, (2) x is initialized and updated after (3) x can't be immutable.
I also often use a lambda expression like above in C++.

So, I want to introduce `do ... end` expression as the syntax sugar of `-> do ... end()`.

```
p := 42
q := 12

x := do
    a := some_process(p, q)
    if a > 0
        ret p * a
    else
        ret some_process(p, a)
    end
end
```

However, there are some concerns for syntax.  It is ambiguous when UFCS invocation without parenthesis when there is no argument.

```
foo.somefunction do
    println("foo!")
end
```

Here, it is parsed as UFCS invocation `foo.somefunction()` with `do ... end` block.  But it also can be interpreted as `foo.somefunction(do ... end)` when `do ... end` expression is employed.
If I simply implement `do ... end` expression following this issue, a parser parses above code as `foo.somefunction(do ... end)` because parser looks for expression for argument before parsing `do ... end` block and is satisfied when the `do ... end` could be parsed as `do ... end` expression.
I want this UFCS invocation to be parsed as `foo.somefunction() do ... end` even if `do ... end` block is introduced.  I think more experiments are needed.

**Comment by rhysd** [comment]
`do ... end` expression makes many syntactic pitfalls because block syntax conflicts as described above.
So I want to introduce `begin ... end` expression instead of it.

```
p := 42
q := 12

x := begin
    a := some_process(p, q)
    if a > 0
        ret p * a
    else
        ret some_process(p, a)
    end
end
```

Simply `do` is replaced with `begin`.  Keyword is different from block syntax.

Then I think I should remove `do ... end` statement by introducint `begin ... end` expression because `do ... end` statement can be written by `begin ... end` expression simply.

For example,  `let ... in` statement with `begin ... end` expression will become like below:

```
let
  x := i + 10
  y := i - 10
in begin
  println(x * y)
end
```

As an optimization, `begin ... end` expression may be compiled to simply-local-scoped statement block when the type is `()` and returned value is not used.  However, I should implement `begin ... end` expression as definition and invocation of lambda expression like above at first.

**Comment by rhysd** [comment]
In addition, I come up with an idea making `let ... in` be expression because `begin ... end` expression is available instead of `do ... end` statement.  I can use it like below:

```
a := [1,2,3]

sum :=
  let
    var s := 0
  in begin
    for e in a
      s += e
    end
    ret s
  end
```

**Comment by rhysd** [comment]
Though it can be written like below,

```
a := [1,2,3]

sum :=
  begin
    var s := 0
    for e in a
      s += e
    end
    ret s
  end
```

I want to introduce `let ... in` expression because separating initialization and the process for it clearly is good idea, I think.  And I could introduce new feature in `let ... in` initialization in the future (e.g. pattern match there).

**Comment by rhysd** [comment]
I've noticed a problem at migrating `let in` from statement to expression.
When `let ... in` get to be an expression, it is hard to introduce new local scope although `let ... in` introduces new variables.

One of the solution is `let ... in` expression invokes lambda expression like below:

- before

```
let
  a := 1
  b := 2
in 1 + 2
```

- after

```
-> do
  a := 1
  b := 2
  ret 1 + 2
end
```

**Comment by rhysd** [comment]
I added `let ... begin ... end` expression.  This is equivalent to `let ... in begin ... end`.  The reason I added it is to avoid nested lambda expression.

```
let
  a := 1
in begin
  println(a)
  println(-a)
end
```

invokes

```
-> do
  a := 1
  -> do
    println(a)
    println(-a)
  end
end
```

So I wanted to reduce nested lambda expression.

```
let
  a := 1
begin
  println(a)
  println(-a)
end
```

invokes

```
-> do
  a := 1
  println(a)
  println(-a)
end
```

At first I implemented handling `let ... in begin ... end` as special case, but I thought it was better to add new (light) syntax to `let` expression and remained `let ... in begin ... end` as it was.
